accepted/00070-akka.md

```diff
@@ -0,0 +1,141 @@
+ - pro: currently Taskomatic runs as `root`, but not all tasks really require `root` privileges. Attack surface can thus be reduced by running code with Tomcat privileges by default
+ - con: Taskomatic tasks that do require root privileges will need some different permissions to run. One way is to use Akka Clustering (described below) with two processes (one `root` and one non-`root`) on the same machine
+
+### Leverage Akka Clustering to offload computation onto other machines
+
+Akka has an optional module called Clustering that allows to transparently move computation across processes (on the same machine or on different machines in a network). It is thus possible to instantiate Actors "remotely", and Akka takes care of serializing, delivering and deserializing Messages to them transparently.
+
+This has been demoed as proof-of-concept [in the HackWeek PR](https://github.com/uyuni-project/uyuni/pull/1903) (commits after February 15), and it is an interesting door to have opened by the work in this PR.
+
+Before considering this for the main codebase, the following aspects should be evaluated:
+
+ - out of all the workloads migrated to Akka, which ones make sense to distribute? In many cases, Uyuni bottlenecks lie in the database or Salt components, and those would not be affected by such a change
+ - does the trade-off make sense at all from a performance perspective? Addding a network layer introduces latency which might end up negating any speedup gained because of the cluster
+ - how can Akka Cluster nodes connect back to the one only Salt and database instances? During HackWeek, a simple SSH tunnel was used for demonstration's sake. Distributing Salt or the database are separate efforts with their own set of trade-offs
+ - is the extra administration overhead worth it? In which cases? Maintaining a cluster is much more work-intensive than maintaining a single VM
+ - how can we deal with upgrades? Akka Clustering has primitives for versioning, but making them work with existing packaging and delivery processes requires some study
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+ - as with all refactorings, we risk introducing regressions at least initially
+ - Akka itself is a complicated artifact. Although external quality seemed very good during HackWeek (no bugs or strange/undocumented behavior was noticed, great documentation) there is always a chance some internal bug hits us, and might be difficult to solve
+ - we get to maintain a packaged version of Akka
+ - performance has to be measured - we expect it to improve but that is not guaranteed
+
+# Alternatives
+[alternatives]: #alternatives
+
+The only considered alternative at this point is to roll our own thread pool model and infrastructure and use it to replace all current implementations. This is expected to be much more difficult and error-prone.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+- Just how far do we feel like going in the steps listed above?
```

Comment:
I would feel confident about a first implementation round covering steps 1-4 mentioned above, but converting Taskomatic to be completely based on Akka seems like a potentially independent effort to better be tackled once the first and easier steps are done and better experience is gathered? It is good to mention the whole plan here, but I could imagine that some technical details are still unknown by now that might require an additional RFC or an extension to this one later. As I understood the prototype does not cover anything related to Taskomatic yet.

Reply:
Step 5 should not be very difficult, considering that some Taskomatic tasks to use the `MessageQueue`, and that means we will have an instance of the `ActorManager` in the Taskomatic process already by step 4.

Step 6 is more uncertain I can agree. Certainly it won't be started until 1-5 are completed, and at that point some research and even a second RFC might be needed. I can make that explicit in the text, moving step 6. to the "Next steps", if you feel it's wiser to do so.

Reply:
Yes, I agree, please clearly separate what you are aiming for in this RFC and what's left for the future.

Reply:
Done